File: easy_1/perfect-number/ruby/v2/perfect_number.rb
# perfect_number.rb

class PerfectNumber
  def self.classify(int)
    raise if int < 1

    sum = self
         .get_factors(int)
         .sum
    return 'perfect' if sum == int
    return 'abundant' if sum > int
    'deficient'
  end

  def self.get_factors(int)
    factors = []
    idx = 1
    limit = int / 2
    loop do
      break if idx > limit

      factors << idx if int % idx == 0
      idx += 1
    end
    factors
  end
end

p PerfectNumber.get_factors(28)

File: easy_1/perfect-number/js/perfect_number.js
class PerfectNumber {
  static classify(num) {
    if (num < 1) throw new Error('Invalid number');

    const factors = PerfectNumber.getFactors(num);
    const sum = factors.reduce( (acc, int) => acc + int);

    if (sum === num) { return "perfect" }
    if (sum > num) { return "abundant" }
    return "deficient";
  }

  static getFactors(num) {
    const limit = Math.floor(num / 2);
    const factors = [];

    for(let idx = 1; idx <= limit; idx += 1) {
      if (num % idx === 0) {
        factors.push(idx)
      }
    }

    return factors;
  }
}

export default PerfectNumber;